catlearn/composition_graph.py

```diff
@@ -302,31 +303,47 @@ def flush(self) -> None:
 
     def arrows(
             self, src: Optional[NodeType] = None,
-            tar: Optional[NodeType] = None
+            tar: Optional[NodeType] = None,
+            arrow_length_range: Tuple[int, Optional[int]] = (0, None),
         ) -> Iterator[CompositeArrow[NodeType, ArrowType]]:
         """
         Get an iterator over all arrows starting at src and ending at tar.
         If source or tar is None, will loop through all possible sources
         and arrows.
         If no existing arrows (or src/tar not in the underlying graph),
-        returns an empty iterator
+        returns an empty iterator.
+        An arrow length range can also be specified. In this case,
+        only arrows with a length in the specified range are returned
         """
-        if src is None and tar is None:
-            # iterate over all edges of graph in this case
-            return iter(self)
+        min_length = arrow_length_range[0]
+        max_length = arrow_length_range[1] if arrow_length_range[1] else inf
+
         if (
                 src is not None and tar is not None
                 and self.graph.has_edge(src, tar)):
             # iterate over all edges from src to tar
-            return (arr.suspend(src, tar) for arr in self.graph[src][tar])
+            # careful to length: the derived arrow has length diminished by 1
```

Comment:
Not sure I follow here. At the very least, I really don't get the `min_length - 1`

Reply:
It is because the arrow as stored as a label of the graph does not contain the first and last node.
So instead of being an arrow with n+1 nodes, n relations, it is stored as an arrow with n nodes, n-1 relations.